server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java

```diff
@@ -162,7 +169,16 @@ public TItem apply(ShardedRequests<TReq, TItem> shardedRequests, Row row, Boolea
 
             RowSourceInfo rowSourceInfo = RowSourceInfo.emptyMarkerOrNewInstance(sourceUri, lineNumber);
             ShardLocation shardLocation = getShardLocation(indexName, id, routing);
+            IndexItem indexItem = (IndexItem) item;
```

Comment:
The `item` can also be a `ShardDeleteRequestItem` which doesn't implement `IndexItem` so this would fail for delete requests. Such we must surround the checker call with `if (item instanceof IndexItem indexItem) {}`

Reply:
Grouper is used only for inserts/`ShardingUpsertExecutor`, `ShardDeleteRequestItem` is used only in `ShardDMLExecutor`. I added an assertion.